apps/remote_control/lib/lexical/remote_control/build/document/compilers/heex.ex

```diff
@@ -0,0 +1,145 @@
+defmodule Lexical.RemoteControl.Build.Document.Compilers.HEEx do
+  @moduledoc """
+  A compiler for .heex files
+  """
+  alias Lexical.Document
+  alias Lexical.Plugin.V1.Diagnostic.Result
+  alias Lexical.RemoteControl.Build
+  alias Lexical.RemoteControl.Build.Document.Compiler
+  alias Lexical.RemoteControl.Build.Document.Compilers
+  require Logger
+
+  @behaviour Compiler
+
+  def recognizes?(%Document{} = document) do
+    Path.extname(document.path) == ".heex"
+  end
+
+  def enabled? do
+    true
+  end
+
+  def compile(%Document{} = document) do
+    with {:ok, _quoted} <- heex_to_quoted(document),
+         {:ok, eex_quoted_ast} <- eval(document) do
+      compile_quoted(document, eex_quoted_ast)
+    end
+  end
+
+  defp compile_quoted(%Document{} = document, quoted) do
+    with {:error, errors} <- Compilers.Quoted.compile(document, quoted, "HEEx") do
+      {:error, reject_undefined_assigns(errors)}
+    end
+  end
+
+  defp heex_to_quoted(%Document{} = document) do
+    try do
+      source = Document.to_string(document)
+
+      opts =
+        [
+          source: source,
+          file: document.path,
+          caller: __ENV__,
+          engine: Phoenix.LiveView.TagEngine,
+          subengine: Phoenix.LiveView.Engine,
+          tag_handler: Phoenix.LiveView.HTMLEngine
+        ]
+
+      quoted = EEx.compile_string(source, opts)
+
+      {:ok, quoted}
+    rescue
+      error ->
+        {:error, [error_to_result(document, error)]}
+    end
+  end
+
+  defp eval(%Document{} = document) do
+    # Evaluating the Html.Engine compiled quoted doesn't report any errors,
+    # so we need to use the original `EEx` to compile it to quoted and evaluate it.
+    quoted_ast =
+      document
+      |> Document.to_string()
+      |> EEx.compile_string(file: document.path)
+
+    result =
+      if Elixir.Features.with_diagnostics?() do
+        eval_quoted_with_diagnostics(quoted_ast, document.path)
+      else
+        eval_quoted(quoted_ast, document.path)
+      end
+
+    case result do
+      {:ok, quoted_ast} ->
+        {:ok, quoted_ast}
+
+      {:exception, exception, stack} ->
+        converted =
+          document
+          |> Build.Error.error_to_diagnostic(exception, stack, quoted_ast)
+          |> Map.put(:source, "HEEx")
+
+        {:error, [converted]}
+
+      {{:ok, quoted_ast}, _} ->
+        # Ignore warnings for now
+        # because they will be handled by `compile_quoted/2`
+        # like: `assign @thing not available in EEx template`
```

Comment:
Maybe instead of ignoring them, I should do something like `reject_undefined_assigns` in a unified place, for all the errors.

Reply:
yes